File: Linear Control Course/Chapter6/Lesson4/Chapter6_Lesson4.py
import numpy as np
import matplotlib.pyplot as plt

# If available, use the python-control package
# pip install control
import control as ctl

# Baseline second-order parameters
zeta = 0.4
wn   = 5.0  # rad/s

# Baseline second-order transfer function: G0(s) = wn^2 / (s^2 + 2 zeta wn s + wn^2)
num0 = [wn**2]
den0 = [1.0, 2.0*zeta*wn, wn**2]
G0 = ctl.TransferFunction(num0, den0)

# Extra pole at s = -a
a = 20.0  # fast pole
kp = a
num_p = [kp * wn**2]
den_p = np.polymul([1.0, a], den0)
Gp = ctl.TransferFunction(num_p, den_p)

# Extra left-half-plane zero at s = -z
z = 2.0
kz = 1.0 / z
num_z = [kz*wn**2, kz*z*wn**2]  # (s + z)*wn^2 / z
den_z = den0
Gz = ctl.TransferFunction(num_z, den_z)

t = np.linspace(0, 4.0, 1000)
t0, y0 = ctl.step_response(G0, T=t)
tp, yp = ctl.step_response(Gp, T=t)
tz, yz = ctl.step_response(Gz, T=t)

plt.figure()
plt.plot(t0, y0, label="Second-order")
plt.plot(tp, yp, label="With extra pole at -a")
plt.plot(tz, yz, label="With zero at -z")
plt.xlabel("Time [s]")
plt.ylabel("Response y(t)")
plt.title("Effect of Additional Pole and Zero on Step Response")
plt.grid(True)
plt.legend()
plt.show()

File: Tutorials/Codes/Linear_Control_Course/Chapter6/Lesson4/Chapter6_Lesson4.m
% Parameters
zeta = 0.4;
wn   = 5.0;

% Baseline second-order
num0 = wn^2;
den0 = [1  2*zeta*wn  wn^2];
G0 = tf(num0, den0);

% Extra pole at s = -a
a  = 20;
kp = a;
num_p = kp * wn^2;
den_p = conv([1 a], den0);
Gp = tf(num_p, den_p);

% Extra zero at s = -z
z  = 2;
kz = 1/z;
num_z = [kz*wn^2  kz*z*wn^2];  % (s + z)*wn^2 / z
den_z = den0;
Gz = tf(num_z, den_z);

% Step responses
t = 0:0.001:4;
[y0, t0] = step(G0, t);
[yp, tp] = step(Gp, t);
[yz, tz] = step(Gz, t);

figure;
plot(t0, y0, 'LineWidth', 1.5); hold on;
plot(tp, yp, 'LineWidth', 1.5);
plot(tz, yz, 'LineWidth', 1.5);
grid on;
xlabel('Time [s]');
ylabel('Response y(t)');
legend('Second-order', 'Extra pole', 'Extra zero');

title('Effect of Additional Pole and Zero on Step Response');

% -------------------------
% Simple Simulink setup (script-driven)
% -------------------------
% You can create a Simulink model programmatically:
%   new_system('extraPoleZeroModel');
%   open_system('extraPoleZeroModel');
% Then insert Transfer Fcn blocks with numerator/denominator corresponding
% to G0, Gp, and Gz, and connect them to a Step block and Scope blocks.
% Robotics System Toolbox can be used to connect these low-order models
% to manipulator or mobile-robot dynamics blocks.
